File: src/main/java/com/joecollins/graphics/screens/generic/SeatsChangingScreen.java
package com.joecollins.graphics.screens.generic;

import static com.joecollins.models.general.PartyResult.NO_RESULT;

import com.joecollins.bindings.Bindable;
import com.joecollins.bindings.BindableList;
import com.joecollins.bindings.Binding;
import com.joecollins.bindings.BindingReceiver;
import com.joecollins.bindings.IndexedBinding;
import com.joecollins.graphics.components.ResultListingFrame;
import com.joecollins.graphics.utils.ColorUtils;
import com.joecollins.graphics.utils.StandardFont;
import com.joecollins.models.general.Party;
import com.joecollins.models.general.PartyResult;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.GridLayout;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.function.UnaryOperator;
import java.util.stream.Collectors;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.border.EmptyBorder;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.ImmutableTriple;
import org.apache.commons.lang3.tuple.Pair;

public class SeatsChangingScreen extends JPanel {

  private SeatsChangingScreen(JLabel title, ResultListingFrame frame) {
    setBackground(Color.WHITE);
    setLayout(new BorderLayout());
    add(title, BorderLayout.NORTH);

    JPanel panel = new JPanel();
    panel.setBackground(Color.WHITE);
    panel.setBorder(new EmptyBorder(5, 5, 5, 5));
    panel.setLayout(new GridLayout(1, 1));
    panel.add(frame);
    add(panel, BorderLayout.CENTER);
  }

  public static <T> Builder<T> of(
      Binding<Map<T, Map<Party, Integer>>> prevResultBinding,
      Binding<Map<T, PartyResult>> currResultBinding,
      Function<T, String> nameFunc,
      Binding<String> headerBinding) {
    return new Builder<>(prevResultBinding, currResultBinding, nameFunc, headerBinding);
  }

  public static class Builder<T> {

    private final BindingReceiver<Map<T, Map<Party, Integer>>> prevResults;
    private final BindingReceiver<Map<T, PartyResult>> currResults;
    private final BindingReceiver<String> header;
    private final Function<T, String> nameFunc;

    private BindingReceiver<Integer> numRows = new BindingReceiver<>(() -> 20);
    private BindingReceiver<Set<T>> seatFilter = new BindingReceiver<>(Binding.fixedBinding(null));

    public Builder(
        Binding<Map<T, Map<Party, Integer>>> prevResultBinding,
        Binding<Map<T, PartyResult>> currResultBinding,
        Function<T, String> nameFunc,
        Binding<String> headerBinding) {
      this.prevResults = new BindingReceiver<>(prevResultBinding);
      this.currResults = new BindingReceiver<>(currResultBinding);
      this.header = new BindingReceiver<>(headerBinding);
      this.nameFunc = nameFunc;
    }

    public Builder<T> withNumRows(Binding<Integer> numRowsBinding) {
      this.numRows = new BindingReceiver<>(numRowsBinding);
      return this;
    }

    public Builder<T> withSeatFilter(Binding<Set<T>> seatFilterBinding) {
      this.seatFilter = new BindingReceiver<>(seatFilterBinding);
      return this;
    }

    public SeatsChangingScreen build(Binding<String> titleBinding) {
      JLabel headerLabel = new JLabel();
      headerLabel.setFont(StandardFont.readBoldFont(32));
      headerLabel.setHorizontalAlignment(JLabel.CENTER);
      headerLabel.setBorder(new EmptyBorder(5, 0, -5, 0));
      titleBinding.bindLegacy(headerLabel::setText);

      Input<T> inputs = new Input<>();
      prevResults.getBinding().bindLegacy(inputs::setPrevResults);
      currResults.getBinding().bindLegacy(inputs::setCurrResults);
      seatFilter.getBinding().bindLegacy(inputs::setSeatFilter);

      BindableList<Entry<T>> entries = new BindableList<>();
      inputs.getResultBinding().bindLegacy(entries::setAll);

      ResultListingFrame frame = new ResultListingFrame();
      frame.setHeaderBinding(header.getBinding());
      frame.setNumRowsBinding(numRows.getBinding());
      frame.setNumItemsBinding(Binding.sizeBinding(entries));
      frame.setTextBinding(IndexedBinding.propertyBinding(entries, e -> nameFunc.apply(e.key)));
      frame.setBorderBinding(IndexedBinding.propertyBinding(entries, e -> e.prevColor));
      frame.setBackgroundBinding(
          IndexedBinding.propertyBinding(entries, e -> e.fill ? e.resultColor : Color.WHITE));
      frame.setForegroundBinding(
          IndexedBinding.propertyBinding(entries, e -> !e.fill ? e.resultColor : Color.WHITE));
      return new SeatsChangingScreen(headerLabel, frame);
    }
  }

  private static class Input<T> extends Bindable<Input<T>, Input.Property> {
    private enum Property {
      PREV,
      CURR,
      FILTER
    }

    private List<Pair<T, Party>> prevResults = List.of();
    private Map<T, PartyResult> currResults = Map.of();
    private Set<T> seatFilter = null;

    public void setPrevResults(Map<T, Map<Party, Integer>> prevResults) {
      this.prevResults =
          prevResults.entrySet().stream()
              .map(
                  e -> {
                    Map<Party, Integer> votes = e.getValue();
                    int total = votes.values().stream().mapToInt(i -> i).sum();
                    List<Integer> topTwo =
                        votes.values().stream()
                            .sorted(Comparator.reverseOrder())
                            .limit(2)
                            .collect(Collectors.toList());
                    return ImmutablePair.of(e, 1.0 * (topTwo.get(0) - topTwo.get(1)) / total);
                  })
              .sorted(Map.Entry.comparingByValue())
              .map(Map.Entry::getKey)
              .map(
                  e ->
                      ImmutablePair.of(
                          e.getKey(),
                          e.getValue().entrySet().stream()
                              .max(Map.Entry.comparingByValue())
                              .map(Map.Entry::getKey)
                              .orElse(null)))
              .collect(Collectors.toList());
      onPropertyRefreshed(Property.PREV);
    }

    public void setCurrResults(Map<T, PartyResult> currResults) {
      this.currResults = currResults;
      onPropertyRefreshed(Property.CURR);
    }

    public void setSeatFilter(Set<T> seatFilter) {
      this.seatFilter = seatFilter;
      onPropertyRefreshed(Property.FILTER);
    }

    public Binding<List<Entry<T>>> getResultBinding() {
      return Binding.propertyBinding(
          this,
          t ->
              t.prevResults.stream()
                  .map(
                      e ->
                          ImmutableTriple.of(
                              e.getLeft(),
                              e.getRight(),
                              t.currResults.getOrDefault(e.getLeft(), NO_RESULT)))
                  .filter(e -> e.getRight() != null)
                  .filter(e -> e.getRight().getParty() != null)
                  .filter(e -> !e.getMiddle().equals(e.getRight().getParty()))
                  .map(
                      e -> {
                        UnaryOperator<Color> colorFunc =
                            (t.seatFilter == null || t.seatFilter.contains(e.getLeft()))
                                ? UnaryOperator.identity()
                                : (c -> ColorUtils.lighten(ColorUtils.lighten(c)));
                        return new Entry<>(
                            e.getLeft(),
                            colorFunc.apply(e.getMiddle().getColor()),
                            colorFunc.apply(e.getRight().getParty().getColor()),
                            e.getRight().isElected());
                      })
                  .collect(Collectors.toList()),
          Property.PREV,
          Property.CURR,
          Property.FILTER);
    }
  }

  private static class Entry<T> {
    private final T key;
    private final Color prevColor;
    private final Color resultColor;
    private final boolean fill;

    private Entry(T key, Color prevColor, Color resultColor, boolean fill) {
      this.key = key;
      this.prevColor = prevColor;
      this.resultColor = resultColor;
      this.fill = fill;
    }
  }
}

File: src/main/java/com/joecollins/graphics/screens/generic/SeatsChangingScreen.kt
package com.joecollins.graphics.screens.generic

import com.joecollins.bindings.Bindable
import com.joecollins.bindings.Binding
import com.joecollins.bindings.BindingReceiver
import com.joecollins.bindings.mapElements
import com.joecollins.graphics.components.ResultListingFrame
import com.joecollins.graphics.utils.ColorUtils.lighten
import com.joecollins.graphics.utils.StandardFont.readBoldFont
import com.joecollins.models.general.Party
import com.joecollins.models.general.PartyResult
import java.awt.BorderLayout
import java.awt.Color
import java.awt.GridLayout
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.border.EmptyBorder

class SeatsChangingScreen private constructor(title: JLabel, frame: ResultListingFrame) : JPanel() {
    class Builder<T>(
        prevResultBinding: Binding<Map<T, Map<Party, Int>>>,
        currResultBinding: Binding<Map<T, PartyResult?>>,
        private val nameFunc: (T) -> String,
        headerBinding: Binding<String?>
    ) {

        private val prevResults: BindingReceiver<Map<T, Map<Party, Int>>> = BindingReceiver(prevResultBinding)
        private val currResults: BindingReceiver<Map<T, PartyResult?>> = BindingReceiver(currResultBinding)
        private val header: BindingReceiver<String?> = BindingReceiver(headerBinding)
        private var numRows = BindingReceiver(Binding.fixedBinding(20))
        private var seatFilter = BindingReceiver<Set<T>?>(Binding.fixedBinding(null))

        fun withNumRows(numRowsBinding: Binding<Int>): Builder<T> {
            numRows = BindingReceiver(numRowsBinding)
            return this
        }

        fun withSeatFilter(seatFilterBinding: Binding<Set<T>?>): Builder<T> {
            seatFilter = BindingReceiver(seatFilterBinding)
            return this
        }

        fun build(titleBinding: Binding<String?>): SeatsChangingScreen {
            val headerLabel = JLabel()
            headerLabel.font = readBoldFont(32)
            headerLabel.horizontalAlignment = JLabel.CENTER
            headerLabel.border = EmptyBorder(5, 0, -5, 0)
            titleBinding.bind { headerLabel.text = it }
            val inputs = Input<T>()
            prevResults.getBinding().bind { inputs.setPrevResults(it) }
            currResults.getBinding().bind { inputs.setCurrResults(it) }
            seatFilter.getBinding().bind { inputs.setSeatFilter(it) }
            val frame = ResultListingFrame(
                headerBinding = header.getBinding(),
                numRowsBinding = numRows.getBinding(),
                itemsBinding = inputs.resultBinding.mapElements {
                    ResultListingFrame.Item(
                        text = nameFunc(it.key),
                        border = it.prevColor,
                        background = (if (it.fill) it.resultColor else Color.WHITE),
                        foreground = (if (!it.fill) it.resultColor else Color.WHITE)
                    )
                }
            )
            return SeatsChangingScreen(headerLabel, frame)
        }
    }

    private class Input<T> : Bindable<Input<T>, Input.Property>() {
        private enum class Property {
            PREV, CURR, FILTER
        }

        private var prevResults: List<Pair<T, Party>> = emptyList()
        private var currResults: Map<T, PartyResult?> = emptyMap()
        private var seatFilter: Set<T>? = null

        fun setPrevResults(prevResults: Map<T, Map<Party, Int>>) {
            this.prevResults = prevResults.entries
                    .asSequence()
                    .map { e: Map.Entry<T, Map<Party, Int>> ->
                        val votes = e.value
                        val total = votes.values.sum()
                        val topTwo = votes.values
                                .sortedDescending()
                                .take(2)
                                .toList()
                        Pair(e, 1.0 * (topTwo[0] - topTwo[1]) / total)
                    }
                    .sortedBy { it.second }
                    .map { it.first }
                    .map { it: Map.Entry<T, Map<Party, Int>> ->
                        Pair(
                                it.key,
                                it.value.entries
                                        .maxByOrNull { it.value }
                                        !!.key)
                    }
                    .toList()
            onPropertyRefreshed(Property.PREV)
        }

        fun setCurrResults(currResults: Map<T, PartyResult?>) {
            this.currResults = currResults
            onPropertyRefreshed(Property.CURR)
        }

        fun setSeatFilter(seatFilter: Set<T>?) {
            this.seatFilter = seatFilter
            onPropertyRefreshed(Property.FILTER)
        }

        val resultBinding: Binding<List<Entry<T>>>
            get() = Binding.propertyBinding(
                    this,
                    { t: Input<T> ->
                        t.prevResults
                                .asSequence()
                                .map {
                                    Triple(
                                            it.first,
                                            it.second,
                                            t.currResults[it.first] ?: PartyResult.NO_RESULT)
                                }
                                .filter { it.third.party != null }
                                .filter { it.second != it.third.party }
                                .map {
                                    val seatFilter = t.seatFilter
                                    val colorFunc = if (seatFilter == null || seatFilter.contains(it.first)) { c: Color -> c } else { c: Color -> lighten(lighten(c)) }
                                    Entry(
                                            it.first,
                                            colorFunc(it.second.color),
                                            colorFunc(it.third.party!!.color),
                                            it.third.isElected)
                                }
                                .toList()
                    },
                    Property.PREV,
                    Property.CURR,
                    Property.FILTER)
    }

    private class Entry<T>(val key: T, val prevColor: Color, val resultColor: Color, val fill: Boolean)
    companion object {
        @JvmStatic fun <T> of(
            prevResultBinding: Binding<Map<T, Map<Party, Int>>>,
            currResultBinding: Binding<Map<T, PartyResult?>>,
            nameFunc: (T) -> String,
            headerBinding: Binding<String?>
        ): Builder<T> {
            return Builder(prevResultBinding, currResultBinding, nameFunc, headerBinding)
        }
    }

    init {
        background = Color.WHITE
        layout = BorderLayout()
        add(title, BorderLayout.NORTH)
        val panel = JPanel()
        panel.background = Color.WHITE
        panel.border = EmptyBorder(5, 5, 5, 5)
        panel.layout = GridLayout(1, 1)
        panel.add(frame)
        add(panel, BorderLayout.CENTER)
    }
}
